### PR TITLE
Move scalar function definitions from scalar.hpp to scalar.cpp

### PR DESCRIPTION
**Pull request by davidwendt**
The `scalar.hpp` is included in over 70 source files in libcudf and many of its function definitions require it to include other headers. This PR moves all function definitions from the `scalar.hpp` to `scalar.cpp` to help minimize the include dependencies and compile time effected by changes to the functions.

This means the templated functions must be defined with all possible types in the .cpp file. This helps reduce the compile parsing and should help simplify the link step since all types are compiled only once. So adding a new type now shows as a compile/link error if it is missing from the `scalar.cpp` definitions.

Overall, this improved compile time by about 2-3 minutes or about 10% for my system running `ninja -j16 cudf` with gcc-9.3, nvcc-11.2.142.

### CODE REVIEW

**Comment by davidwendt** [comment]
The recent addition of the `list_scalar` introduced the `column.hpp` header to `scalar.hpp`.
I experimented with using [pimpl](https://herbsutter.com/gotw/_100/) to remove the header but the compile time did not change for me.

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/scalar/scalar.cpp [hunk 1]
```diff
@@ -56,4 +95,352 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
...
+template class fixed_width_scalar<timestamp_D>;
+template class fixed_width_scalar<timestamp_s>;
+template class fixed_width_scalar<timestamp_ms>;
+template class fixed_width_scalar<timestamp_us>;
+template class fixed_width_scalar<timestamp_ns>;
+template class fixed_width_scalar<duration_D>;
+template class fixed_width_scalar<duration_s>;
+template class fixed_width_scalar<duration_ms>;
+template class fixed_width_scalar<duration_us>;
+template class fixed_width_scalar<duration_ns>;
```
Comment (ttnghia) [inline cpp/src/scalar/scalar.cpp:247]: I would like to use a macro to iterate through all types in a loop at compile time? Something similar to this:
```
#define INSTANTIATE(Type, ...)  \
    template class fixed_width_scalar(Type); \
    INSTANTIATE(__VA_ARGS__)

INSTANTIATE(int8_t, int16_t, ........................, duration_ns)
```

File: cpp/src/scalar/scalar.cpp [hunk 2]
```diff
@@ -56,4 +95,352 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
...
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr);
+template timestamp_scalar<timestamp_ns>::timestamp_scalar(duration_ns const& value,
+                                                          bool is_valid,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr);
+template timestamp_scalar<timestamp_ns>::timestamp_scalar(int64_t const& value,
+                                                          bool is_valid,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr);
```
Comment (ttnghia) [inline cpp/src/scalar/scalar.cpp:432]: Such long explicit template instantiation can also be shortened by my recommendation above.

File: cpp/src/scalar/scalar.cpp [hunk 3]
```diff
@@ -56,4 +95,352 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
...
+  return _data.data();
+}
+
+template <typename T>
+fixed_width_scalar<T>::operator value_type() const
+{
+  return this->value(rmm::cuda_stream_default);
+}
+
+template class fixed_width_scalar<bool>;
```
Comment (rgsl888prabhu) [inline cpp/src/scalar/scalar.cpp:227]: Better to add comment to update whenever a new type is introduced.

File: cpp/src/scalar/scalar.cpp (hunk 1 above)
Comment (davidwendt) [reply cpp/src/scalar/scalar.cpp:255]: I disagree with this recommendation. 
I think the code here is simpler to understand and maintain as is.

File: cpp/src/scalar/scalar.cpp (hunk 2 above)
Comment (davidwendt) [reply cpp/src/scalar/scalar.cpp]: I think listing the parameters here make this more difficult to update. Removing the names did not seem to help. 
So I created a macro for this case so only one place needs to change if the parameter or definition changes.

**Review by rgsl888prabhu — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

File: cpp/src/scalar/scalar.cpp [hunk 4]
```diff
@@ -56,4 +95,339 @@ std::string string_scalar::to_string(rmm::cuda_stream_view stream) const
...
+  return _data.data();
+}
+
+template <typename T>
+typename fixed_point_scalar<T>::rep_type const* fixed_point_scalar<T>::data() const
+{
+  return _data.data();
+}
+
+template class fixed_point_scalar<numeric::decimal32>;
```
Comment (harrism) [inline cpp/src/scalar/scalar.cpp:167]: Comment here like the ones below saying "These are the valid types of fixed-point scalars"?

**Review by harrism — APPROVED** [review]
Great optimization. One tiny comment suggestion.

**Review by jrhemstad — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**